BringYourOwnBoard.md

```diff
@@ -0,0 +1,157 @@
+
+For instance you can look at this snippet: 
+```u-boot
+fdt addr ${fdt_addr} && fdt get value bootargs /chosen bootargs
+fatload mmc 0:1 ${kernel_addr_r} @@KERNEL_IMAGETYPE@@
+
+if test ! -e @@BOOT_MEDIA@@ 0:1 uboot.env; then saveenv; fi;
+
+if env exists rpipart;
+then 
+  if test "${have_updated}" = 1; then
+    if test "${boot_attempts}" = "3";
+    then 
+      # do a reversion based upon rpipart
+      if test "${rpipart}" = "2"; then
+        # Revert to three
+        setenv rpipart 3
+        setenv boot_attempts 0
+      else
+        if test "${rpipart}" = "3"; then
+          # Revert to two
+          setenv rpipart 2
+          setenv boot_attempts 0
+        fi
+      fi
+
+    fi
+  fi
+  
+  setenv bootargs "${bootargs} root=/dev/mmcblk0p${rpipart}"; 
+fi
+
+setenv boot_attempts boot_attempts + 1
+@@KERNEL_BOOTCMD@@ ${kernel_addr_r} - ${fdt_addr}
+```
+
+Within the [recipes-core] layer we create three partitions and mount them in the `fstab` : `/dev/mmcblk0p2` (the 'A' partition), `/dev/mmcblk0p3` (the 'B' partition), and `/dev/mmcblk0p4` (the "adu" partition). When an update occurs we write the partition number to boot into next to the u-boot environment section setup using `fw-env.config` which is then used by this u-boot script to determine what partition to load into on the next boot. 
+
+For a production device the core idea of using a flag to say what to load into is completely acceptable, however there are some safety checks you should add if you want your device to be secure and durable. 
+
+First you should also write the hash of the image to be booted into. Then u-boot should hash the partition you're attempting to write into (at least include a CRC checksum) and determine if the partition is what you expect. This prevents direct flash injection attacks and keeps your device from running unsafe code. 
+
+Second you want to add a method for handling a kernel panic. That is if you've completed an update and somehow the partition you're booting into isn't bootable you have a way to fall back from the B side to the A side or vice versa. You can accomplish that using u-boot flags and some logic around what to do when. This is essential to preventing security breaches and making sure your devices have healthy fallbacks. 
+
+Lastly you should look into how your partitions are setup. We only include three partitions + the bootloader in our proof of concept. We use two for updates and one for the Device Update data to be persisted between the two sectors. You may have additional data you need persisted across updates. This would be a place to include some of the boot logic to check whether these sectors are secure or have been messed with since the last boot. Effectively just do secure boot but not secure boot. 
```

Comment:
Is there any suggestions we can provide as a starting point beyond what's here? I'm thinking about something like a link to relevant u-boot documentation, not anything we'd directly support.